gix-worktree/src/diff.rs

```diff
@@ -0,0 +1,292 @@
+    /// Note that this doesn't necessarily mean that the *content* of the file changed
+    /// see [`FileStatus`] for details
+    Modified {
+        /// How the file was modified exactly
+        modification: FileModification,
+        /// Whether this (changed) file also has an unresolved merge conflict
+        conflict: bool,
+    },
+    /// There are unresolved merge conflicts for this file
+    /// but it has not changed on disk
+    Conflict,
+}
+
+/// Computes the changes between the index and the worktree
+pub fn compare_to_index<'a: 'b, 'b>(
+    index: &'a index::State,
+    // TODO: use worktree cache instead
+    worktree: &'b Path,
+    capabilities: &'b fs::Capabilities,
+) -> impl Iterator<Item = Result<Change<'a>, Error>> + 'b {
+    // TODO: parallel with rayon
+    index.entries().iter().filter_map(|index_entry| {
+        let conflict = match index_entry.stage() {
+            0 => false,
+            1 => true,
+            _ => return None,
+        };
+        let git_path = index_entry.path(index);
+        if index_entry.flags.intersects(
+            index::entry::Flags::UPTODATE
+                | index::entry::Flags::SKIP_WORKTREE
+                | index::entry::Flags::ASSUME_VALID
+                | index::entry::Flags::FSMONITOR_VALID,
+        ) {
+            return None;
+        }
+
+        let path = if let Ok(path) = path::try_from_bstr(git_path) {
+            path
+        } else {
+            return Some(Err(Error::IllformedUtf8 {
+                path: git_path.to_owned(),
+            }));
+        };
+
+        let worktree_path = worktree.join(path);
+        let metadata = match worktree_path.symlink_metadata() {
+            // TODO: check if any parent directory is a symlink
+            // we need to use fs::Cache for that
+            Ok(metadata) if metadata.is_dir() => {
+                // index entries are normally only for files/symlinks
+                // if a file turned into a directory it was removed
+                // the only exception here are submodules which are
+                // part of the index despite being directories
+                //
+                // TODO: submodules:
+                //   if entry.mode.contains(Mode::COMMIT) &&
+                //     resolve_gitlink_ref(ce->name, "HEAD", &sub))
+                return Some(Ok(Change {
+                    kind: ChangeKind::Removed,
+                    index_entry,
+                    worktree_path,
+                    fstat: Some(metadata),
+                }));
+            }
+            Ok(metdata) => metdata,
+            Err(err) if err.kind() == ErrorKind::NotFound => {
+                return Some(Ok(Change {
+                    kind: ChangeKind::Removed,
+                    index_entry,
+                    worktree_path,
+                    fstat: None,
+                }))
+            }
+            Err(err) => {
+                // TODO: strict mode?
```

Comment:
An interesting question - it's easy to be strict in Rust even though the C version of this might be quite the opposite. What does `git` do in this case? If it's more lenient, we'd need that too I guess, otherwise it could be done on popular demand.

Reply:
This comment is kind of redudant now.  Previously I had a visitor pattern based approach where this would abort the entire function (basically the entire function returned one result). With the filter map we just yield the error to the caller and they can decide whether to abort on the first error, ignore them or deal with them otherwise.

Git handles errors inconsistently and differently in different parts of the codebase. In the function I am emulating here error just lead to the file being add to the diffq as modified but later when the diff is actually performed they might actually lead to an abort. I think with this kind of iterator style API letting the user decide how to deal with errors works decently well

Reply:
I generally love iterator-style APIs as well as they enable incremental and interruptible computation. Usually these are harder to implement and parallelize though and because of the latter it might be better to move away from it like you suggested.